src/planner/rules/range.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 RisingLight Project Authors. Licensed under Apache-2.0.
+
+//! Range filter.
+
+use std::ops::Bound;
+
+use super::*;
+use crate::catalog::ColumnRefId;
+use crate::storage::KeyRange;
+
+/// The data type of range analysis.
+///
+/// If Some, the expression is a range condition.
+///
+/// ```text
+/// a = 1
+/// a > 1
+/// a <= 1
+/// -1 < a < 1
+/// ```
+pub type RangeCondition = Option<(ColumnRefId, KeyRange)>;
+
+/// Returns all columns involved in the node.
+pub fn analyze_range(egraph: &EGraph, enode: &Expr) -> RangeCondition {
+    use Expr::*;
+    let column = |i: &Id| {
+        egraph[*i].nodes.iter().find_map(|e| match e {
+            Expr::Column(c) => Some(*c),
+            _ => None,
+        })
+    };
+    let range = |i: &Id| egraph[*i].data.range.as_ref();
+    let constant = |i: &Id| egraph[*i].data.constant.as_ref();
+    match enode {
+        Eq([a, b]) | Gt([a, b]) | GtEq([a, b]) | Lt([a, b]) | LtEq([a, b]) => {
+            // normalize `v op k` to `k op v`
+            let (k, v, enode) = if let (Some(v), Some(k)) = (constant(a), column(b)) {
+                let revnode = match enode {
+                    Eq(_) => Eq([*b, *a]),
+                    Gt(_) => Lt([*b, *a]),
+                    GtEq(_) => LtEq([*b, *a]),
+                    Lt(_) => Gt([*b, *a]),
+                    LtEq(_) => GtEq([*b, *a]),
+                    _ => unreachable!(),
+                };
+                (k, v, revnode)
+            } else if let (Some(k), Some(v)) = (column(a), constant(b)) {
+                (k, v, enode.clone())
+            } else {
+                return None;
+            };
+            let start = match enode {
+                Eq(_) | GtEq(_) => Bound::Included(v.clone()),
+                Gt(_) => Bound::Excluded(v.clone()),
+                Lt(_) | LtEq(_) => Bound::Unbounded,
+                _ => unreachable!(),
+            };
+            let end = match enode {
+                Eq(_) | LtEq(_) => Bound::Included(v.clone()),
+                Lt(_) => Bound::Excluded(v.clone()),
+                Gt(_) | GtEq(_) => Bound::Unbounded,
+                _ => unreachable!(),
+            };
+            Some((k, KeyRange { start, end }))
+        }
+        And([a, b]) => {
+            let (ka, ra) = range(a)?;
+            let (kb, rb) = range(b)?;
+            if ka != kb {
+                return None;
+            }
+            // if both a and b have bound at start or end, return None
+            let start = match (&ra.start, &rb.start) {
+                (Bound::Unbounded, s) | (s, Bound::Unbounded) => s.clone(),
+                _ => return None,
+            };
+            let end = match (&ra.end, &rb.end) {
+                (Bound::Unbounded, s) | (s, Bound::Unbounded) => s.clone(),
+                _ => return None,
+            };
```

Comment:
what if both range from one side are bound? (i.e., a < 3 && a < 5) In this case we should not return a `None` here if we don't have filter executor above scan-filter?

Reply:
Yes it will return a `None`. I didn't merge the bounds here because I thought it would be a bit complicated. For these cases I expected them to be handled in rewrite rules later. For example:
```
(and (< ?a ?v1) (< ?a ?v2)) => (< ?a (min ?v1 ?v2))
```